### PR TITLE
Implement MSC3531: Hiding messages during moderation

### PR DESCRIPTION
**Pull request by Yoric**
Matrix-js-sdk-level support for hiding messages.

This is my first patch in matrix-js-sdk, please accept my apologies for anything I have done wrong.

MSC: https://github.com/matrix-org/matrix-doc/pull/3531

Signed-off-by: David Teller <davidt@element.io>

---
This PR currently has no changelog labels, so will not be included in changelogs.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.

### CODE REVIEW

File: src/models/event.ts [hunk 1]
```diff
@@ -154,13 +154,27 @@ export interface IDecryptOptions {
...
+}
+
 export class MatrixEvent extends EventEmitter {
     private pushActions: IActionsObject = null;
     private _replacingEvent: MatrixEvent = null;
     private _localRedactionEvent: MatrixEvent = null;
     private _isCancelled = false;
     private clearEvent?: IClearEvent;
 
+    /* Message hiding, as specified by MSC3515.
```
Comment (turt2live) [inline src/models/event.ts:168]: wrong MSC number? I suggest links to avoid people having to convert between these manually.

File: src/models/event.ts [hunk 2]
```diff
@@ -934,6 +948,28 @@ export class MatrixEvent extends EventEmitter {
         this.event.unsigned.redacted_because = redactionEvent.event as IEvent;
     }
 
+    /**
+     * Apply a MSC3531 event to hide/unhide the message.
```
Comment (turt2live) [inline src/models/event.ts:952]: links please

File: src/models/event.ts [hunk 3]
```diff
@@ -934,6 +948,28 @@ export class MatrixEvent extends EventEmitter {
...
+     *
+     * @param visibilityEvent event causing the hide/unhide change.
+     * @return true if the event caused a change, false if it should be ignored.
+     */
+    public applyVisibilityChange(visible: boolean, reason?: string): boolean {
+        let change;
```
Comment (turt2live) [inline src/models/event.ts:958]: types on `change` please, or update the ladder to early return.

File: src/models/event.ts [hunk 4]
```diff
@@ -934,6 +948,28 @@ export class MatrixEvent extends EventEmitter {
         this.event.unsigned.redacted_because = redactionEvent.event as IEvent;
     }
 
+    /**
+     * Apply a MSC3531 event to hide/unhide the message.
+     *
+     * @param visibilityEvent event causing the hide/unhide change.
```
Comment (turt2live) [inline src/models/event.ts:954]: This isn't a parameter

File: src/models/event.ts [hunk 5]
```diff
@@ -1003,6 +1039,14 @@ export class MatrixEvent extends EventEmitter {
         return this.getType() === "m.room.redaction";
     }
 
+    public isVisibilityChange(): boolean {
```
Comment (turt2live) [inline src/models/event.ts:1042]: Visibility can mean all sorts of things, so I'd suggest we name this to avoid confusion with directory and room visibility.
```suggestion
    public isEventVisibilityChange(): boolean {
```

File: src/models/event.ts [hunk 6]
```diff
@@ -1003,6 +1039,14 @@ export class MatrixEvent extends EventEmitter {
...
+        const relation = this.getRelation();
+        if (!relation) {
+            return false;
+        }
+        return relation.rel_type == "m.visibility";
```
Comment (turt2live) [inline src/models/event.ts:1047]: This cannot and should not be using the `m.*` namespace yet. Please use your unstable prefix, ideally through the `UnstableValue` class.

File: src/models/room.ts [hunk 7]
```diff
@@ -151,6 +156,14 @@ export class Room extends EventEmitter {
      */
     public threads = new Map<string, Thread>();
 
+    /**
+     * A mapping of eventId to the latest visibility change to apply
+     * to the event.
+     *
+     * @experimental
```
Comment (turt2live) [inline src/models/room.ts:163]: Would be good to link to the MSC here

File: src/models/room.ts [hunk 8]
```diff
@@ -1365,6 +1378,71 @@ export class Room extends EventEmitter {
             // this may be needed to trigger an update.
         }
 
+        if (event.isVisibilityChange()) {
+            // Apply MSC3531-style visibility changes.
+            (() => {
```
Comment (turt2live) [inline src/models/room.ts:1383]: Why is this a closure? (comments please)

File: src/models/room.ts [hunk 9]
```diff
@@ -1365,6 +1378,71 @@ export class Room extends EventEmitter {
...
+                // Perform sanity checks on the event.
+                const relation = event.getRelation();
+                const originalEventId = relation.event_id;
+                let visible: boolean;
+                switch (relation["visibility"]) {
```
Comment (turt2live) [inline src/models/room.ts:1388]: The MSC requires an unstable prefix for this field.

**Review by turt2live — CHANGES_REQUESTED** [review]
As a first pass there's mostly spec-related workflow concerns on this, but the change overall seems sensible. I've also left a review on the MSC itself which might ultimately impact the review of the actual visibility change code.

**Comment by Yoric** [comment]
Apologies, I realized too late that I was missing a big part of the patch to handle redactions. Added.

File: src/models/room.ts (hunk 8 above)
Comment (Yoric) [reply src/models/room.ts]: Moved to a method, it's more readable that way.

File: src/models/event.ts [hunk 10]
```diff
@@ -1003,6 +1065,80 @@ export class MatrixEvent extends EventEmitter {
...
+            UNSTABLE_MSC3531_REASON_FIELD.unstable]) {
+            if (!key) {
+                continue;
+            }
+            const reason = content[key];
+            if (!reason && typeof reason == "string") {
+                return reason;
+            }
+        }
+        return null;
```
Comment (turt2live) [inline src/models/event.ts:1139]:
```suggestion
        return UNSTABLE_MSC3531_REASON_FIELD.findIn(content);
```

File: src/models/room.ts [hunk 11]
```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
...
+
+    /**
+     * When we receive a new visibility change event:
+     *
+     * - store this visibility change alongside the timeline, in case we
+     *   later need to apply it to an event that we haven't received yet;
+     * - if we have already received the event whose visibility has changed,
+     *   patch it to reflect the visibility change and inform listeners.
+     */
+    private applyNewVisibilityChangeEvent(event: MatrixEvent): void {
```
Comment (turt2live) [inline src/models/room.ts:2300]: from before: we should be consistent in the naming. If we're calling this an EventVisibility, then we should always use "EventVisibility".

Applies to the other functions as well.

File: src/models/room.ts [hunk 12]
```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
...
+            case "hidden":
+                visible = false;
+                break;
+            case "visible":
+                visible = true;
+                break;
+            default:
+                // Event is ill-formed.
+                return;
+        }
```
Comment (turt2live) [inline src/models/room.ts:2320]: This can be shortened:

```suggestion
        visible = event.getEventVisibilityChange();
        if (!visible && visible !== false) return; // event is ill-formed
```

File: src/models/room.ts [hunk 13]
```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
...
+        const reason = visible ? event.getContent()["reason"] : null;
+        if (reason != null && reason !== undefined && typeof reason != "string") {
```
Comment (turt2live) [inline src/models/room.ts:2322]:
```suggestion
        if (reason && typeof reason != "string") {
```

File: src/models/room.ts [hunk 14]
```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
...
+            // Event is ill-formed.
+            return;
+        }
+
+        // Ignore visibility change events that are not emitted by moderators.
+        const powerLevelsEvents = this.currentState.getStateEvents(EventType.RoomPowerLevels, "");
+        const powerLevels = powerLevelsEvents && powerLevelsEvents.getContent();
+        if (!powerLevels || powerLevels["org.matrix.msc3531.visibility"] < event.sender.powerLevel) {
```
Comment (turt2live) [inline src/models/room.ts:2330]: unstable value for the namespace please :)

File: src/models/room.ts [hunk 15]
```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
...
+            // 2 visibility changes, so linear scans through `visibilityChanges`
+            // should be fast. It might be possible to DoS this feature, though.
+            let index = visibilityChangesOnOriginalEvent.length - 1;
+            for (; index >= 0; --index) {
+                const target = visibilityChangesOnOriginalEvent[index];
+                if (target.getTs() < event.getTs()) {
+                    break;
+                }
+            }
+            if (index == -1) {
```
Comment (turt2live) [inline src/models/room.ts:2353]:
```suggestion
            if (index === -1) {
```

File: src/models/room.ts [hunk 16]
```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
...
+        // we may need to apply the visibility change at a later date.
+
+        const visibilityChangesOnOriginalEvent = this.visibilityChanges.get(originalEventId);
+        if (visibilityChangesOnOriginalEvent) {
+            // It would be tempting to simply erase the latest visibility change
+            // but we need to record all of the changes in case the latest change
+            // is ever redacted.
+            // In practice, it should be unusual for an event to have more than
+            // 2 visibility changes, so linear scans through `visibilityChanges`
+            // should be fast. It might be possible to DoS this feature, though.
```
Comment (turt2live) [inline src/models/room.ts:2345]: We can always apply a limit to the loop, and we probably should if it's highlighted as an attack here.

File: src/models/room.ts [hunk 17]
```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
...
+        const relation = event.getRelation();
+        const originalEventId = relation.event_id;
+        const visibilityChangesOnOriginalEvent = this.visibilityChanges.get(originalEventId);
+        if (!visibilityChangesOnOriginalEvent) {
+            // No visibility changes on the original event.
+            // In particular, this change event was not recorded,
+            // most likely because it was ill-formed.
+            return;
+        }
+        const index = visibilityChangesOnOriginalEvent.findIndex(change => change.getId() == event.getId());
```
Comment (turt2live) [inline src/models/room.ts:2388]:
```suggestion
        const index = visibilityChangesOnOriginalEvent.findIndex(change => change.getId() === event.getId());
```

File: src/models/room.ts [hunk 18]
```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
...
+        if (index == -1) {
```
Comment (turt2live) [inline src/models/room.ts:2389]:
```suggestion
        if (index === -1) {
```

Comment (turt2live) [reply src/models/room.ts:2389]: this sort of comment applies throughout - going to stop posting about it for this review :)

File: src/models/room.ts [hunk 19]
```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
...
+    /**
+     * When we receive an event whose visibility has been altered by
+     * a (more recent) visibility change event, patch the event in
+     * place so that clients now not to display it.
+     *
+     * @param event Any matrix event. If this event has at least one a
+     * pending visibility change event, apply the latest visibility
+     * change event.
+     */
+    private applyPendingVisibilityChanges(event: MatrixEvent): void {
```
Comment (turt2live) [inline src/models/room.ts:2447]: same comment as the other function: EventVisibility please :)

**Review by turt2live — COMMENTED** [review]
overall this is looking fine - just a few more nitpicks, mostly around the project structure

File: src/models/room.ts (hunk 12 above)
Comment (Yoric) [reply src/models/room.ts:2320]: Not sure I understand how you go from a string to a boolean.

File: src/models/room.ts (hunk 13 above)
Comment (Yoric) [reply src/models/room.ts:2322]: That would mean that if `reason == ""`, we behave as if `reason` was `null`? That makes sense.

File: src/models/room.ts (hunk 16 above)
Comment (Yoric) [reply src/models/room.ts:2345]: How should we react when we reach the limit?

Comment (turt2live) [reply src/models/room.ts:2354]: I'd just truncate. If someone is flipping visibility more than 50 times, they kinda deserve whatever state we ended up on.

File: src/@types/event.ts [hunk 20]
```diff
@@ -180,26 +180,32 @@ export const UNSTABLE_ELEMENT_FUNCTIONAL_USERS = new UnstableValue(
 /**
  * A rel_type that affects visibility of a message,
  * as per https://github.com/matrix-org/matrix-doc/pull/3531
+ *
+ * @experimental
  */
-export const UNSTABLE_MSC3531_VISIBILITY_CHANGE_REL_TYPE = new UnstableValue(
+export const MSC3531_VISIBILITY_CHANGE_REL_TYPE = new UnstableValue(
```
Comment (turt2live) [inline src/@types/event.ts:186]: sorry, also meant to say that the MSC number should be dropped from the constant's name as well. Also, `EVENT` should be somewhere in there to line up with all the other EventVisibility things.

File: src/models/room.ts (hunk 12 above)
Comment (turt2live) [reply src/models/room.ts]: per internal: I'm not really seeing the value of the string, so would change this to be a boolean.

File: src/models/room.ts (hunk 13 above)
Comment (turt2live) [reply src/models/room.ts]: there are entire swathes of code in our projects which rely on empty strings being falsy, so I wouldn't worry about it.

File: src/models/event.ts [hunk 21]
```diff
@@ -923,6 +983,53 @@ export class MatrixEvent extends EventEmitter {
...
+     *   changing the reason (if it was hidden).
+     * @param visibilityEvent event holding a hide/unhide payload, or nothing
+     *   if the event is being reset to its original visibility (presumably
+     *   by a visibility event being redacted).
+     */
+    public applyVisibilityEvent(visibilityChange?: IVisibilityChange): void {
+        const visible = visibilityChange ? visibilityChange.visible : true;
+        const reason = visibilityChange ? visibilityChange.reason : null;
+        let change = false;
+        if (this.visibility.visible != visibilityChange.visible) {
```
Comment (turt2live) [inline src/models/event.ts:1001]: triple equals throughout please.

File: src/models/room-state.ts [hunk 22]
```diff
@@ -607,13 +608,15 @@ export class RoomState extends EventEmitter {
...
-     * @param {string} eventType The type of event to test
+     * @param {string} eventType The type of event to test.
      * @param {string} userId The user ID of the user to test permission for
      * @return {boolean} true if the given user ID should be permitted to send
      *                        the given type of event into this room,
      *                        according to the room's state.
      */
-    public maySendEvent(eventType: EventType | string, userId: string): boolean {
+    public maySendEvent<S extends string, T extends string>(
+        eventType: EventType | string | UnstableValue<S, T>,
```
Comment (turt2live) [inline src/models/room-state.ts:618]: These functions should not be taking `UnstableValue` directly. The caller will need to call the function for both types if it wants to.

This should also remove the iterator requirement on `UnstableValue` - something that shouldn't be present there.

File: yarn.lock [hunk 23]
```diff
@@ -1210,9 +1210,9 @@
     "@types/yargs" "^15.0.0"
     chalk "^4.0.0"
 
-"@matrix-org/olm@https://gitlab.matrix.org/api/v4/projects/27/packages/npm/@matrix-org/olm/-/@matrix-org/olm-3.2.3.tgz":
-  version "3.2.3"
-  resolved "https://gitlab.matrix.org/api/v4/projects/27/packages/npm/@matrix-org/olm/-/@matrix-org/olm-3.2.3.tgz#cc332fdd25c08ef0e40f4d33fc3f822a0f98b6f4"
+"@matrix-org/olm@https://gitlab.matrix.org/api/v4/projects/27/packages/npm/@matrix-org/olm/-/@matrix-org/olm-3.2.7.tgz":
```
Comment (turt2live) [inline yarn.lock:1213]: not sure what happened here, but this seems unrelated.

**Review by turt2live — CHANGES_REQUESTED** [review]
almost there :)

There's also MSC feedback which will need addressing ftr. Those aspects are not covered in code review.

File: src/@types/event.ts [hunk 24]
```diff
@@ -177,6 +177,16 @@ export const UNSTABLE_ELEMENT_FUNCTIONAL_USERS = new UnstableValue(
...
 
+/**
+ * A type of message that affects visibility of a message,
+ * as per https://github.com/matrix-org/matrix-doc/pull/3531
+ *
+ * @experimental
+ */
+export const EVENT_VISIBILITY_CHANGE_TYPE = new UnstableValue(
+    "org.matrix.msc3531.visibility",
+    "m.visibility");
```
Comment (turt2live) [inline src/@types/event.ts:188]: These arguments are backwards

File: src/models/room-state.ts (hunk 22 above)
Comment (Yoric) [reply src/models/room-state.ts:618]: > These functions should not be taking `UnstableValue` directly. The caller will need to call the function for both types if it wants to.

Why? It feels to me like this is both a common operation when prototyping events and something error-prone (and slower) to reimplement manually.

> This should also remove the iterator requirement on UnstableValue - something that shouldn't be present there.

I don't understand that remark, either.

Comment (t3chguy) [reply src/models/room-state.ts:618]: The logic would end up duplicated anyway as the user may have permission to send one but not the other so the called needs to know which one in the case of knowing whether it may have to send the unstable one because thats the only one it has perms to send to, if that's even a thing it wishes to do.

Comment (Yoric) [reply src/models/room-state.ts:618]: > 

How's that different from the code I wrote that we're discussing?

Comment (turt2live) [reply src/models/room-state.ts:618]: Behaviourally it's fine for your PR, but it isn't necessarily safe for other implementations or features. It's very possible that a feature down the line needs explicitly one or the other, and this signature would introduce a potential false sense of security to a developer.

Comment (Yoric) [reply src/models/room-state.ts]: Ok, I've changed the implementation.

File: src/models/room-state.ts [hunk 25]
```diff
@@ -605,27 +605,27 @@ export class RoomState extends EventEmitter {
...
-     *                        according to the room's state.
-     */
+      * Returns true if the given user ID has permission to send a normal
+      * event of type `eventType` into this room.
+      * @param {string} eventType The type of event to test
+      * @param {string} userId The user ID of the user to test permission for
+      * @return {boolean} true if the given user ID should be permitted to send
+      *                        the given type of event into this room,
+      *                        according to the room's state.
+      */
```
Comment (turt2live) [inline src/models/room-state.ts:615]: all of the jsdoc comments appear to have gotten mangled by the revert

File: src/models/room.ts [hunk 26]
```diff
@@ -2254,6 +2311,152 @@ export class Room extends EventEmitter {
...
+            // The event is ill-formed.
+            return;
+        }
+
+        // Ignore visibility change events that are not emitted by moderators.
+        const userId = event.getSender();
+        if (!userId) {
+            return;
+        }
+        if (!EVENT_VISIBILITY_CHANGE_TYPE.some(name => this.currentState.maySendStateEvent(name, userId))) {
```
Comment (turt2live) [inline src/models/room.ts:2335]: I'm not really convinced that we need an iterator on this - we can just call `maySendStateEvent` twice. It'd be a bit different if we were supporting more than two namespaces or the classes had a purpose outside of detecting presence within an object.

Or in other words: the usage of `some` feels backwards here. A namespaced value is intended to retrieve values from an object/array, not process an entity to check compliance. 

Storing two variables and checking either is more than fine, I think. The power level check uses local cache so should have no performance impact.

File: yarn.lock [hunk 27]
```diff
@@ -1210,9 +1210,9 @@
     "@types/yargs" "^15.0.0"
     chalk "^4.0.0"
 
-"@matrix-org/olm@https://gitlab.matrix.org/api/v4/projects/27/packages/npm/@matrix-org/olm/-/@matrix-org/olm-3.2.3.tgz":
-  version "3.2.3"
-  resolved "https://gitlab.matrix.org/api/v4/projects/27/packages/npm/@matrix-org/olm/-/@matrix-org/olm-3.2.3.tgz#cc332fdd25c08ef0e40f4d33fc3f822a0f98b6f4"
+"@matrix-org/olm@https://gitlab.matrix.org/api/v4/projects/27/packages/npm/@matrix-org/olm/-/@matrix-org/olm-3.2.8.tgz":
```
Comment (turt2live) [inline yarn.lock:1213]: This seems unrelated.

**Review by turt2live — COMMENTED** [review]
one more step, and then we're there :)

File: src/models/room.ts (hunk 26 above)
Comment (Yoric) [reply src/models/room.ts:2335]: It strikes me as error-prone, but I'll stop insisting.

File: yarn.lock (hunk 27 above)
Comment (Yoric) [reply yarn.lock:1213]: Gasp. I'm almost sure that I have reverted that change several times already. Not sure what's going on.

**Review by turt2live — APPROVED** [review]
alright, let's send it.

**Comment by turt2live** [comment]
... after you fix merge conflicts and the build :D (try merging develop - should fix both, hopefully)

**Comment by turt2live** [comment]
(please don't force push onto reviewed branches - it invalidates review, thus making it required for re-review. In this case, the changes appear fine)
